SHA3.cs

```diff
@@ -0,0 +1,191 @@
+    public static ValueTask<int> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, Memory<byte> destination, CancellationToken cancellationToken = default);
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+}
+
+public partial class HMACSHA3_384 : HMAC {
+    public const int HashSizeInBits = 384;
+    public const int HashSizeInBytes = 48;
+
+    public HMACSHA3_384();
+    public HMACSHA3_384(byte[] key);
+
+    public static bool IsSupported { get; } // New
+
+    public static byte[] HashData(byte[] key, byte[] source);
+    public static byte[] HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source);
+    public static int HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination);
+    public static bool TryHashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten);
+
+    public static byte[] HashData(byte[] key, Stream source);
+    public static byte[] HashData(ReadOnlySpan<byte> key, Stream source);
+    public static int HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination);
+
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+    public static ValueTask<int> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, Memory<byte> destination, CancellationToken cancellationToken = default);
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+}
+
+public partial class HMACSHA3_512 : HMAC {
+    public const int HashSizeInBits = 512;
+    public const int HashSizeInBytes = 64;
+
+    public HMACSHA3_512();
+    public HMACSHA3_512(byte[] key);
+
+    public static bool IsSupported { get; } // New
+
+    public static byte[] HashData(byte[] key, byte[] source);
+    public static byte[] HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source);
+    public static int HashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination);
+    public static bool TryHashData(ReadOnlySpan<byte> key, ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten);
+
+    public static byte[] HashData(byte[] key, Stream source);
+    public static byte[] HashData(ReadOnlySpan<byte> key, Stream source);
+    public static int HashData(ReadOnlySpan<byte> key, Stream source, Span<byte> destination);
+
+    public static ValueTask<byte[]> HashDataAsync(byte[] key, Stream source, CancellationToken cancellationToken = default);
+    public static ValueTask<int> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, Memory<byte> destination, CancellationToken cancellationToken = default);
+    public static ValueTask<byte[]> HashDataAsync(ReadOnlyMemory<byte> key, Stream source, CancellationToken cancellationToken = default);
+}
+
+public abstract partial class Shake : IDisposable {
```

Comment:
I'm not convinced this base class necessary. But I threw it out as a suggestion. To give us a bit of a safety net on this, the class will have an internal constructor so that other developers may not derive from it, only refer to it by type.